### PR TITLE
Implement an optimized merkle trie

### PR DESCRIPTION
**Pull request by ultratwo**
This pull requests adds the optimized Merkle Trie. This faster implementation can be used for the state interface by calling `ethereum.optimized.monkey_patch_optimized_state()` during startup.

There is currently one regression. Equality comparisons between optimized `Trie`/`State` objects do not work (you have to call `root`/`state_root`.

#### Testing

The code in `optimized/test_trie.py` achieves 100% branch coverage on the core logic (excluding a few impossible cases).

I have not added any other tests for the optimized logic. You can run the test suite using the optimized state implementation by adding
```python
import ethereum.optimized

ethereum.optimized.monkey_patch_optimized_state()
```
to `src/ethereum/__init__.py`.

#### Performance
These tests were done on a trie containing 1 million random keys using cpython.

Any operation on the unoptimized trie: 66s
Adding 60 keys on the optimized trie: 3.7ms (62μs per key)

It is unlikely that performance will be a problem on tries that fit in memory.

#### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

**Comment by SamWilsn** [comment]
Two quick questions (haven't started reviewing yet):

 - How would you feel about moving this out of `src/ethereum` so that it doesn't get included in the rendered documentation (ex. `src/ethereum_state`)?
 - Should we manually implement the `__eq__` method for the relevant classes?

**Comment by ultratwo** [comment]
> * How would you feel about moving this out of `src/ethereum` so that it doesn't get included in the rendered documentation (ex. `src/ethereum_state`)?

This sounds like a good idea, I just didn't know where to put it. If you tell me where you think it should go, I'll move it there.

> * Should we manually implement the `__eq__` method for the relevant classes?

I'll push this.

**Review by SamWilsn — COMMENTED** [review]
- I think we can put it in `src/ethereum_optimize` or something like that. I don't care too much about the name.
 - How do we cope with hardforks in the optimized implementation? Should we pass the module into the monkey patch function? Patch them all in a loop?

File: src/ethereum/optimized/state.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
+# type: ignore
```
[the remark this answers is not in the document]
Comment (ultratwo) [reply src/ethereum/optimized/state.py:1]: `state.py` calls lots of functions that take a `frontier.Trie` with an `optimized.Trie` and generally undermines the typesystem with aggressive duck typing. Getting it to typecheck seems hopeless.

File: src/ethereum/optimized/__init__.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import ethereum.optimized.trie as optimized_trie
+
+optimized_state_patches = {
+    "Trie": optimized_trie.Trie,
+    "root": optimized_trie.root,
+    "trie_get": optimized_trie.trie_get,
+    "trie_set": optimized_trie.trie_set,
+    "State": optimized_state.State,  # type: ignore
+    "set_storage": optimized_state.set_storage,  # type: ignore
+    "get_storage": optimized_state.get_storage,  # type: ignore
```
[the remark this answers is not in the document]
Comment (ultratwo) [reply src/ethereum/optimized/__init__.py:30]: The error is `src/ethereum/optimized/__init__.py:28: error: Module has no attribute "State"` and is caused by disabling typechecking in `optimized/state.py`.

**Review by SamWilsn — APPROVED** [review]

**Pull request merged**